File: Structural/Decorator/src/examplejava/datasource/FileDataSource.java
package examplejava.datasource;

import java.io.*;

//Class, which implements straight file reading and writing
public class FileDataSource implements DataSource {
    private String name;

    public FileDataSource(String name) {
        this.name = name;
    }

    @Override
    public void writeData(String data) {
        File file = new File(name);
        try (OutputStream fileOutputStream = new FileOutputStream(file)) {
            fileOutputStream.write(data.getBytes(), 0, data.length());
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
    }

    @Override
    public String readData() {
        char[] buffer = null;
        File file = new File(name);
        try (FileReader fileReader = new FileReader(name)) {
            buffer = new char[(int) file.length()];
            fileReader.read(buffer);
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        return null;
    }
}

File: Structural/Decorator/src/examplekotlin/datasource/FileDataSource.kt
package examplekotlin.datasource

import java.io.*

class FileDataSource constructor(val name: String) : DataSource {
    override fun writeData(data: String) {
        val file = File(name)
        FileOutputStream(file).use { writer ->
            writer.write(data.toByteArray(), 0, data.length)
        }
    }

    override fun readData() :String?{
        val buffer: CharArray? = null
        val file = File(name)
        FileReader(file).use { reader ->
            reader.read(buffer)
        }
        return null
    }
}